### PR TITLE
Add ESM2 PEFT recipe

### PR DESCRIPTION
**Pull request by balvisio**
### Description

This PR adds a recipe to perform LoRA fine-tuning to the Evo2 model. It provides support for DDP and sequence packing. It also contains a file `infer.py` that shows how to do inference from a fine-tuned checkpoint. The PR contains the datasets that were used to train the models. Eventually we can convert them into HF datasets.

It does not add support for FSDP or FP8 yet.

#### Usage

```bash
cd bionemo-recipes/recipes/evo2_peft_te/
python train_lora_ddp.py
```

For more information on usage see the README.

### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.

- [ciflow:skip](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:skip) - Skip all CI tests for this PR
- [ciflow:notebooks](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:notebooks) - Run Jupyter notebooks execution tests for bionemo2
- [ciflow:slow](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:slow) - Run slow single GPU integration tests marked as @pytest.mark.slow for bionemo2
- [ciflow:all](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all) - Run all tests (unit tests, slow tests, and notebooks) for bionemo2. This label can be used to enforce running tests for all bionemo2.
- [ciflow:all-recipes](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all-recipes) - Run tests for all recipes (under bionemo-recipes). This label can be used to enforce running tests for all recipes.

Unit tests marked as `@pytest.mark.multi_gpu` or `@pytest.mark.distributed` are not run in the PR pipeline.

For more details, see [CONTRIBUTING](CONTRIBUTING.md)

> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

#### Triggering Code Rabbit AI Review

To trigger a code review from code rabbit, comment on a pull request with one of these commands:

- @coderabbitai review - Triggers a standard review
- @coderabbitai full review - Triggers a comprehensive review

See https://docs.coderabbit.ai/reference/review-commands for a full list of commands.

### Pre-submit Checklist

- [x] I have tested these changes locally
- [x] I have updated the documentation accordingly
- [x] I have added/updated tests as needed
- [x] All existing tests pass successfully

## Release Notes

**New Features**
- Added Parameter-Efficient Fine-Tuning (LoRA) support for ESM2 models, enabling efficient adaptation with reduced memory footprint
- Introduced convolutional token classification head as an alternative architecture for sequence labeling tasks
- New ESM2 PEFT recipe with complete training and inference pipelines, including distributed training support

**Documentation**
- Added comprehensive guides for training, inference, datasets, and Docker deployment
- Provided Hydra configuration examples for quick-start scenarios

**Chores**
- Updated PEFT dependency to specialized version for enhanced functionality
- Enhanced distributed training infrastructure with performance monitoring and logging

### CODE REVIEW

File: bionemo-recipes/recipes/esm2_peft_te/scheduler.py
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_peft_te/scheduler.py:1]: rather than symlinking, can you make these hard copies and add them to `./ci/scripts/check_copied_files.py`?

File: bionemo-recipes/recipes/esm2_peft_te/data/porter6_train_dataset_55k.parquet
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_peft_te/data/porter6_train_dataset_55k.parquet:1]: this is too big to store in the repo... can we find someway to have this on the huggingface hub?

File: bionemo-recipes/recipes/esm2_peft_te/example_nv_esm2_t6_8M_UR50D_peft_checkpoint/adapter_model.safetensors
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_peft_te/example_nv_esm2_t6_8M_UR50D_peft_checkpoint/adapter_model.safetensors:1]: i'm not sure we want to include this directory in the repo

File: bionemo-recipes/recipes/esm2_peft_te/utils.py
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_peft_te/utils.py:1]: could this go in dataset.py? I'm just particularly against files called "utils" 😄

File: bionemo-recipes/recipes/esm2_peft_te/dataset.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+from torch.utils.data import DataLoader, DistributedSampler
+from transformers import (
+    AutoTokenizer,
+    DataCollatorForTokenClassification,
+    DataCollatorWithFlattening,
+)
+
+from collator import TokenPackingDataset
+from distributed_config import DistributedConfig
+from utils import SS3_LABEL2ID, SS8_LABEL2ID
```
[the remark this answers is not in the document]
Comment (balvisio) [reply bionemo-recipes/recipes/esm2_peft_te/dataset.py]: no. i specifically want to use the transformers versio

File: bionemo-recipes/recipes/esm2_peft_te/infer.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        args.model_tag,  # original model tag
+        config=config,
+        trust_remote_code=True,
+    )
+
+    # Load PEFT adapters on top
+    peft_model = PeftModel.from_pretrained(base_model, args.peft_model_config_dir)
+    peft_model = peft_model.to("cuda").eval()
+
+    tokenizer = AutoTokenizer.from_pretrained("nvidia/esm2_t48_15B_UR50D")
```
[the remark this answers is not in the document]
Comment (balvisio) [reply bionemo-recipes/recipes/esm2_peft_te/infer.py:111]: in all cases we want to use the nvidia/* tokenizer. they are all the same

[the remark this answers is not in the document]
Comment (balvisio) [reply bionemo-recipes/recipes/esm2_peft_te/infer.py:111]: it's fine. we always want to use the nvidia/* tokenizer

File: bionemo-recipes/recipes/esm2_peft_te/train_lora_convnet.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+"""Demonstration of LoRA fine-tuning of ESM-2 with Transformer Engine and PEFT."""
+
+import logging
+from pathlib import Path
+
+import hydra
+import peft
+import torch
+from modeling_esm_te import NVEsmForConvTokenClassification
```
[the remark this answers is not in the document]
Comment (balvisio) [reply bionemo-recipes/recipes/esm2_peft_te/train_lora_convnet.py]: fixed

File: bionemo-recipes/recipes/esm2_peft_te/example_nv_esm2_t6_8M_UR50D_peft_checkpoint/adapter_model.safetensors
Comment (balvisio) [reply bionemo-recipes/recipes/esm2_peft_te/example_nv_esm2_t6_8M_UR50D_peft_checkpoint/adapter_model.safetensors:1]: deleted

File: bionemo-recipes/recipes/esm2_peft_te/utils.py
Comment (balvisio) [reply bionemo-recipes/recipes/esm2_peft_te/utils.py:1]: fixed

File: bionemo-recipes/recipes/esm2_peft_te/data/porter6_train_dataset_55k.parquet
Comment (balvisio) [reply bionemo-recipes/recipes/esm2_peft_te/data/porter6_train_dataset_55k.parquet:1]: deleted

File: .github/workflows/unit-tests-recipes.yml [hunk 4]
```diff
@@ -155,6 +155,9 @@ jobs:
         with:
           sparse-checkout: "${{ matrix.recipe.dir }}"
           sparse-checkout-cone-mode: false
+      - name: Include symlink targets for esm2_peft_te
+        if: ${{ matrix.recipe.dir == 'bionemo-recipes/recipes/esm2_peft_te' }}
+        run: git -c safe.directory=/__w/bionemo-framework/bionemo-framework sparse-checkout add bionemo-recipes/recipes/esm2_native_te
```
Comment (pstjohn) [inline .github/workflows/unit-tests-recipes.yml]: remove

File: bionemo-recipes/models/esm2/pyproject.toml [hunk 5]
```diff
@@ -14,7 +14,7 @@ dependencies = [
     "jinja2",
     "megatron-fsdp",
     "omegaconf",
-    "peft",
+    "peft @ git+https://github.com/balvisio/peft.git@support-te-lora",
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/pyproject.toml]: let's leave the test xfailing in esm2 (with ToT `peft`) until/unless it's merged upstream

File: bionemo-recipes/recipes/esm2_native_te/checkpoint.py
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/checkpoint.py:1]: these checkpoint files don't have to be identical, i'd revert this

File: bionemo-recipes/recipes/esm2_peft_te/tests/peft_test_dataset.csv
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_peft_te/tests/peft_test_dataset.csv:1]: probably easier just to write this at test time through something like

```python
@pytest.fixture
def test_csv_file(tmp_path):
    path = tmp_path / "test_dataset.csv"
    path.write_text(textwrap.dedent("""\
        Sequence,Secondary_structure
        LAGVSERTIDPKQNFYMHWC,HIGEBST~HIGEBST~HIGE
        """)
    return path
```

File: ci/scripts/check_copied_files.py [hunk 6]
```diff
@@ -57,6 +57,18 @@
     "bionemo-recipes/models/esm2/tests/common": [
         "bionemo-recipes/models/llama3/tests/common",
     ],
+    "bionemo-recipes/recipes/esm2_native_te/collator.py": [
+        "bionemo-recipes/recipes/esm2_peft_te/collator.py",
+    ],
```
Comment (pstjohn) [inline ci/scripts/check_copied_files.py]: add this to the existing `collator.py` list above

File: ci/scripts/check_copied_files.py [hunk 7]
```diff
@@ -57,6 +57,18 @@
...
+    "bionemo-recipes/recipes/esm2_native_te/checkpoint.py": [
+        "bionemo-recipes/recipes/esm2_peft_te/checkpoint.py",
+    ],
+    "bionemo-recipes/recipes/esm2_native_te/distributed_config.py": [
+        "bionemo-recipes/recipes/esm2_peft_te/distributed_config.py",
+    ],
+    "bionemo-recipes/recipes/esm2_native_te/scheduler.py": [
+        "bionemo-recipes/recipes/esm2_peft_te/scheduler.py",
+    ],
```
Comment (pstjohn) [inline ci/scripts/check_copied_files.py]: i wouldn't bother enforcing these to stay in sync

File: .github/workflows/unit-tests-recipes.yml (hunk 4 above)
Comment (balvisio) [reply .github/workflows/unit-tests-recipes.yml]: done

File: bionemo-recipes/models/esm2/pyproject.toml (hunk 5 above)
Comment (balvisio) [reply bionemo-recipes/models/esm2/pyproject.toml]: done

File: bionemo-recipes/recipes/esm2_peft_te/tests/peft_test_dataset.csv
Comment (balvisio) [reply bionemo-recipes/recipes/esm2_peft_te/tests/peft_test_dataset.csv:1]: done

File: bionemo-recipes/recipes/esm2_peft_te/infer.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+def _batched_inference(
+    model,
+    tokenizer,
+    records,
+    batch_size: int,
+    max_seq_length: int,
+    stride: int,
+    infer_overflowing_aas: bool,
+    device: str = "cuda",
+) -> tuple[list[str], list[int]]:
```
[the remark this answers is not in the document]
Comment (balvisio) [reply bionemo-recipes/recipes/esm2_peft_te/infer.py:36]: no need

File: ci/scripts/check_copied_files.py (hunk 6 above)
Comment (balvisio) [reply ci/scripts/check_copied_files.py]: done

File: ci/scripts/check_copied_files.py (hunk 7 above)
Comment (balvisio) [reply ci/scripts/check_copied_files.py]: done

File: bionemo-recipes/recipes/esm2_peft_te/tests/test_infer.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+import peft
+import pytest
+import torch
+from transformers import AutoConfig, AutoModelForTokenClassification, AutoTokenizer
+
+from dataset import SS3_ID2LABEL, SS3_LABEL2ID
+from infer import _batched_inference
+
+
+requires_cuda = pytest.mark.skipif(not torch.cuda.is_available(), reason="Test requires CUDA")
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_peft_te/tests/test_infer.py]: you can remove this; coderabbit and genAI tend to put these in, but all our tests require CUDA 😆
